### PR TITLE
[DS][x/n] Allow SchedulingConditions to target specific dependency keys

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This is a long-standing request. It's often useful to be able to detect the status of some specific set of dependency keys, rather than needing to treat all dependencies identically.

This creates a single structure for defining and resolving these sorts of collections, but does not export it as a top-level user-facing concept, as that feels a bit like jumping the gun.

This allows users to explicitly select any keys they want to select for, as well as explicitly ignore some specific keys, as either workflow is potentially useful. An explicit choice was made to not error when an "invalid" selection (i.e. one that references keys that are not dependencies) is made, as this would be pretty annoying (imo).

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21854?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#21854** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21854?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#21853** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21853?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21844** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21844?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21839** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21839?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/parent_newer_condition.py [hunk 1]
```diff
@@ -6,15 +6,25 @@
...
 class ParentNewerCondition(SchedulingCondition):
+    dep_selection: DepKeySelection = DepKeySelection.all()
+
     @property
     def description(self) -> str:
-        return "At least one parent has been updated more recently than the candidate."
+        description = "At least one parent has been updated more recently than the candidate"
+        if not self.dep_selection.is_all:
+            description += f" within selection: {self.dep_selection.description}"
+        return description
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/parent_newer_condition.py:23]: nit: could be one-liner

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py [hunk 2]
```diff
@@ -103,22 +106,55 @@ def __invert__(self) -> "NotAssetCondition":
         return NotAssetCondition(operand=self)
 
     @staticmethod
-    def any_deps_match(condition: "SchedulingCondition") -> "AnyDepsCondition":
+    def any_deps_match(
+        condition: "SchedulingCondition",
+        allow_keys: Optional[AbstractSet[AssetKey]] = None,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py:111]: Violates https://github.com/dagster-io/internal/discussions/9719

Also some future yet-to-be-written smells around "Parameter Clumping". Let's figure out a better structure here.

**Review by schrockn — CHANGES_REQUESTED** [review]
We need to figure out a way to make this compose better IMO

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/parent_newer_condition.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/parent_newer_condition.py]: Maybe missing something dumb, but I'm not seeing a clean way to make this a oneliner, e.g.:

```python
return "All deps" + (f"within _selection: {self.dep_selection}" if self.dep_selection else "")
```

would work, but it feels a fair bit harder to read to me. Open to suggestions though

**Comment by OwenKephart** [comment]
@schrockn Updated this to just work off of regular-old `AssetSelection`s, and created a common interface thing so we don't have to implement `get_dep_keys()` multiple times.

**Pull request closed without merging**